### ISSUE TITLE
[FEATURE REQUEST] Allow the JavaScript package manager to be configurable

### ISSUE THREAD

**Issue by JakeChampion** [fastly/cli#729]
**Is your feature request related to a problem? Please describe.**
I have a project which uses yarn but when I run fastly compute serve fastly runs npm install 
https://github.com/fastly/cli/blob/903492bfbfa4bbdfa1cd4c36e71ad58a87f9f7f1/pkg/commands/compute/language_javascript.go#LL46C11-L46C11
https://github.com/fastly/cli/blob/903492bfbfa4bbdfa1cd4c36e71ad58a87f9f7f1/pkg/commands/compute/language_javascript.go#L62

**Describe the solution you'd like**
I would like to be able to configure the fastly cli with regards to javascript package manager used as it is currently it is hardcoded to npm

**Describe alternatives you've considered**
An alternative would be to not run any installation step as part of the fastly cli, instead allowing each project to decide whether to run that themselves within the fastly.toml.scripts.build script

**Comment by harmony7** [comment]
maybe possible to use

https://nodejs.org/api/packages.html#packagemanager

**Comment by rkrv** [comment]
+1 on this issue

**Comment by peterjmag** [comment]
Hi all! Just a heads up that the docs still indicate that Yarn is supported via a custom build command:

![Screen Shot 2023-01-17 at 14 20 01](https://user-images.githubusercontent.com/352105/212909430-fb609437-bc67-4c9b-928f-542a718fca2f.png)

https://developer.fastly.com/reference/cli/compute/build/#examples

And this seems to imply that the custom build command should be respected by `compute serve`:

![Screen Shot 2023-01-17 at 14 18 31](https://user-images.githubusercontent.com/352105/212909103-ee480783-8a9f-47fd-8824-9cf2b14a5547.png)

https://developer.fastly.com/reference/cli/compute/serve/#examples

Given that, should this issue be labeled as a bug rather than a feature request?

**Comment by Integralist** [comment]
Ultimately, we should be supporting both yarn and npm so I'll create an internal ticket to track this (especially considering we originally supported yarn, but from my understanding this was unfortunately dropped by mistake after a large internal refactoring).

**Comment by Integralist** [comment]
So a quick grep through the code base we are using these `npm` commands... 

- `npm exec ...`
- `npm init`
- `npm run ...`
- `npm install --save-dev`
- `npm install --global npm@latest`
- `npm list --json --depth 0`

I'm not a JavaScript dev and so I'm not familiar with the npm/yarn toolchain.

Is anyone able to advise on `yarn` equivalents to these commands?

From what I can see `yarn` has the following commands which match by 'name' but maybe not by 'behaviour'...

- `yarn exec`
- `yarn init`
- `yarn run`
- `yarn install`
- `yarn list`

So I'll dig into those and see how they behave in comparison to `npm`.

**Comment by peterjmag** [comment]
@Integralist Most commands are equivalent (though I'm not too familiar with `npm exec` or `yarn exec` myself). Here's a helpful reference: https://gist.github.com/jonlabelle/c082700c1c249d986faecbd5abf7d65b

The biggest difference is `npm install` — Yarn uses `yarn add`, which has slightly different semantics:

- `npm install [package] --save-dev` => `yarn add [package] --dev`
- `npm install --global [package]` => `yarn global add [package]`

I believe the parsing of `package-name@version` is also the same between tools, so that part shouldn't need any translation/adaptation.

**Comment by peterjmag** [comment]
Oof yeah sorry about that @Integralist — I'd forgotten about the whole Yarn 1 vs Yarn 2+ thing, because we're stuck on Yarn 1.x for most of our stuff internally.

**Comment by Integralist** [comment]
@peterjmag That's OK. 

Interesting to know you're on v1 as the Fastly CLI is going to want to support the latest release of Yarn (or at least v2+) rather than the "classic" version (as it refers to it) because v1 just behaves so differently to the other releases, and having to support multiple interfaces for a single tool is probably too much of a maintenance overhead.

**Comment by peterjmag** [comment]
Yep that's totally fair, especially given that v1 has been in maintenance mode for 3 years now:

![Screen Shot 2023-01-23 at 17 27 17](https://user-images.githubusercontent.com/352105/214094428-57f869e5-3027-4454-afe0-d6194dad404c.png)

([via](https://classic.yarnpkg.com/lang/en/docs/install/))

**Comment by rkrv** [comment]
Hey Mark,

Thank you for the quick turnaround on this feature!

> Validate if you had node and npm installed.
   Validate if your package.json had the right dependencies configured.
   Execute an npm install for you and check the associated dependency constraints were met.

For my team, we do the validation/installation internally in our build/dev scripts and use some options in package.json, so we don't really need those things out of the box with `fastly-cli`.

I also wonder if some of that could be replaced with plain old shell scripts that don't rely on npm/yarn and just warn the user if something is amiss but not block the execution (every JS environment and setup is unique and beautiful on its own).

**Comment by Integralist** [comment]
Thanks @rkrv for the feedback! Very much appreciated 🙂 

re:

> I also wonder if some of that could be replaced with plain old shell scripts

Exactly what `[scripts.build]` was hoping to achieve, but having it pre-configured with what was needed to produce a wasm binary but to allow a user to change it to whatever they want (e.g. `npm run my-own-thing`) and then the user could run their build/serve commands with `--skip-verification` to avoid any CLI errors.

So I think the current CLI workflow already enables users to take control (the key bit being to remember to use `--skip-verification`).

But even if users just want to use what the CLI provides out-of-the-box (e.g. npm in this case) the logic for handling validation in the CLI is _very complex_ and understandably so considering it's trying to negotiate its way around an unknown landscape (i.e. the user's local environment) and all the problems and caveats that come with doing that. 

So that's why the internal conversation has just begun, because we're now realising the current implementation isn't maintainable/sustainable long term.

**Comment by Integralist** [comment]
👋🏻 

This is a quick update to let everyone following this issue know that the topic of whether we keep the 'user environment' validation within the Fastly CLI has been discussed and we've decided to remove that logic from the CLI as it will help to reduce the friction for users who want to switch the tooling used, as well as reducing the general confusion caused for users when there _are_ environment validation errors, and lastly, it reduces the maintenance overhead for contributors to the CLI code base as the logic flow becomes much simpler.

So, with that in mind and as already [mentioned earlier](https://github.com/fastly/cli/issues/729#issuecomment-1402096830), as the CLI can support yarn currently with some manual steps I'll close off this issue and I'll schedule the work internally to remove the user validation from the Fastly CLI.

Thank you everyone for your input and feedback. It was very much appreciated.
